packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -627,16 +650,68 @@ export function allocateChildren(vnode: VCustomElement, vm: VM) {
             );
         }
     }
+
+    // If any of the children being allocated are VFragments, we remove the text delimiters and flatten all immediate
+    // children VFragments to avoid them interfering with default slot behavior.
+    const allocatedChildren = flattenFragmentsInChildren(children);
+    vnode.children = allocatedChildren;
+    vm.aChildren = allocatedChildren;
+
     if (shadowMode === ShadowMode.Synthetic || renderMode === RenderMode.Light) {
         // slow path
-        allocateInSlot(vm, children, vnode.owner);
+        allocateInSlot(vm, allocatedChildren, vnode.owner);
         // save the allocated children in case this vnode is reused.
-        vnode.aChildren = children;
+        vnode.aChildren = allocatedChildren;
         // every child vnode is now allocated, and the host should receive none directly, it receives them via the shadow!
         vnode.children = EmptyArray;
     }
 }
 
+/**
+ * Flattens the contents of all VFragments in an array of VNodes, removes the text delimiters on those VFragments, and
+ * marks the resulting children array as dynamic.
+ *
+ * With the delimiters removed, the contents must be treated as dynamic for them to rerender correctly.
+ *
+ * This function is used for slotted VFragments to avoid the text delimiters interfering with slotting functionality.
+ */
+function flattenFragmentsInChildren(children: VNodes): VNodes {
+    const flattenedChildren: VNodes = [];
+    const nodeStack: VNodes = [];
```

Comment:
nit: At first, I was wondering what the purpose of the nodeStack is. May be add a comment to the method and say the method will recursively flatten the (nested) fragments. I also wonder if this piece can be rewritten elegantly using a recursion.

Reply:
I'll add a better comment for this part.

This was actually originally written using recursion. Nolan suggested switching to an iterative traversal to avoid the extra perf overhead of creating lots of throwaway objects/arrays. Conversation for reference: https://github.com/salesforce/lwc/pull/3140#discussion_r1044806776

Reply:
Yes it was my fault to recommend not using recursion. FWIW I believe typically using your own stack is faster than recursion (although it's definitely a micro-optimization thing).

Reply:
Ah sorry I didn't mean to point fingers, just provide some background context. I think your suggestion made sense as a micro-optimization, my implementation just needed better documentation.

Reply:
No worries, I didn't take it that way. 🙂